경북대 BE_유우석 Step3 - 위시리스트 구현

milk-stone opened the pull request:
안녕하십니까 멘토님 Step3 기능을 뒤늦게라도 구현하여 리뷰 요청 드립니다!

제가 지금 학교에서 지원해주는 프로그램을 들으러 미국에 와있는 상태인데, 어제 프레젠테이션 일정이랑 겹쳐서 한국 시간대 기준 토요일 23:59까지 완료하지 못하였습니다.. 현재 한국 시각 기준 오전 4:00 인 걸 알고 있지만, 뒤늦게라도 작성하여 리뷰 요청 드립니다..

늘 좋은 피드백 제공해주셔서 감사합니다!

milk-stone committed:
docs: Step 3 구현할 기능 정리

milk-stone committed:
feat: Wish 엔티티 구현 및 schema.sql에 테이블 생성 쿼리 작성

milk-stone committed:
feat: 위시리스트 API를 작성할 때 필요한 DTO 작성

milk-stone committed:
feat: Wish를 찾지 못했을 때의 커스텀 예외 생성 및 글로벌 예외 처리기에 설정

milk-stone committed:
feat: 위시리스트 API Repository 메서드 작성

milk-stone committed:
feat: 위시리스트 API Service 메서드 작성

milk-stone committed:
feat: 위시리스트 API Controller 메서드 작성

milk-stone committed:
feat: 위시리스트 수량 수정 API - 수량 0으로 변경 시 삭제되도록 로직 추가

milk-stone committed:
fix: 로그인 실패, 토큰 만료시 에러코드로 403 대신 401 반환하도록 수정

Rok93 commented on README.md:
```diff
@@ -55,3 +55,23 @@
+
+| URL | 메서드 | 기능 | 설명                                    |
+| :--- | :--- | :--- |:--------------------------------------|
+| /api/wishlist | GET | 다건 위시리스트 조회 | 전체 위시리스트 목록을 조회                       |
+| /api/wishlist | POST | 위시리스트 생성 | 위시리스트에 상품 추가                          |
+| /api/wishlist/{id} | PUT | 위시리스트 수정 | 특정 위시리스트 항목의 상품 수량을 수정 (수량이 0이 되면 삭제) |
+| /api/wishlist/{id} | DELETE | 위시리스트 삭제 | 특정 상품을 위시리스트에서 삭제                     |
```
이번 단계 요구 사항에는 `위시리스트 수정 API`를 구현하라는 요구사항은 없는 것 같아요. 🤔 

<img width="525" height="160" alt="image" src="https://github.com/user-attachments/assets/45492cc6-fa0d-4c7e-8ee6-7ae84b5b1637" />

milk-stone replied:
<img width="456" height="189" alt="image" src="https://github.com/user-attachments/assets/2a937137-cf61-471b-97bd-af48965c966a" />

이 시퀀스 다이어그램을 보고 수정 API를 작성하려고 했던 거긴 한데, 수량이 0이 되는걸 수정 없이 어떻게 구현해야할지 조금 고민해보아야 할 것 같네요.! 생각 나면 또 댓글 달겠습니다!

Rok93 replied:
오호... 시퀀스 다이아그램에서는 수정에 대한 언급이 있었군요? 🤔

[API 명세](https://edu.nextstep.camp/s/UH6H8Iba/ls/TxIElkRG) 페이지에 보시면 아래와 같이 적혀있는데요. 
이런 관점에서는 우석님이 누락된 API를 잘 찾아주신게 아닐까 싶어요! 👍💯
꼼꼼하게 잘 봐주셨네요. 🤩 

> API 명세는 아래와 같다. 클라이언트에 따라 이러한 API가 모두 필요할 수도 있고 필요하지 않을 수도 있다. 또는 누락된 API가 있을 수도 있다.

Rok93 commented on README.md:
```diff
@@ -55,3 +55,23 @@
+| :--- | :--- | :--- |:--------------------------------------|
+| /api/wishlist | GET | 다건 위시리스트 조회 | 전체 위시리스트 목록을 조회                       |
+| /api/wishlist | POST | 위시리스트 생성 | 위시리스트에 상품 추가                          |
+| /api/wishlist/{id} | PUT | 위시리스트 수정 | 특정 위시리스트 항목의 상품 수량을 수정 (수량이 0이 되면 삭제) |
```
어짜피 수정은 요구사항에 없어 제거(?)될 것 같지만 
수량이 0이면 삭제로 연결되기보다는, 0으로 수정할 수 없도록 유효성 검증을 해주는 것이 좋다고 생각해요. 
API도 하나의 역할만하도록하는 것이 추후 유지보수하기 수월할 것 같아요! 😃

milk-stone replied:
API가 하나의 역할만 하도록 하도록 수정해보겠습니다! 좋은 관점인 것 같습니다

Rok93 commented on src/main/java/gift/domain/wish/Wish.java:
```diff
@@ -0,0 +1,45 @@
+    private Long productId;
+    private int quantity;
+
+    protected Wish() {
+    }
```
기본 생성자가 없어도될 것 같은데, 필요한 이유가 있을까요? 🤔

milk-stone replied:
없어도 될 것 같다 라는 표현이 정확하게 일치하네요.. 이유가 없는 코드는 작성하지 않는게 더 좋은 습관 같습니다.

추가했던 이유는 얕은 지식이지만, 대부분의 프레임워크는 리플렉션(Reflection)이라는 기술을 기반으로 동작한다고 알고 있고, 리플렉션이 작동하려면 인자 없는 기본 생성자가 있어야 하는 걸로 알고 있어서 작성했었습니다.

없어도 작동되니까 일단은 지워보겠습니다!

Rok93 replied:
습관적인 코드가 은근히 많죠 😃
getter, setter도 사실 우석님이 말씀하신 것과 비슷한 케이스이고요. 

하지만 이런 불필요한 코드들은 결국 잘못 쓰일 위험이 생길 수 있기 때문에 정말 필요한 상황이라면 어쩔 수 없지만 필요 없는 상황이라면 애초에 실수할 여지를 없애두는 것이 좋다고 생각해요. 🤗

Rok93 replied:
습관적인 코드가 은근히 많죠 😃
getter, setter도 사실 우석님이 말씀하신 것과 비슷한 케이스이고요. 

하지만 이런 불필요한 코드들은 결국 잘못 쓰일 위험이 생길 수 있기 때문에 정말 필요한 상황이라면 어쩔 수 없지만 필요 없는 상황이라면 애초에 실수할 여지를 없애두는 것이 좋다고 생각해요. 🤗

Rok93 commented on src/main/java/gift/domain/wish/service/WishService.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    public void updateWish(Long id, WishUpdateRequest req, String accessToken) {
+        Member member = authService.getMemberByToken(accessToken).orElseThrow(()->new TokenExpiredException(accessToken));
```
이 부분은 코드 자동정렬이 안된 것 같은데요. 🤔 
습관적으로 자동정렬을 진행해주시거나 혹은 IDE 셋팅으로 코드 변경을 저장할 때마다 자동으로 정렬 기능이 동작하도록 셋팅을 해주시면 좋을 것 같습니다. 🙏

milk-stone replied:
자동 정렬 기능이 있는지 처음 알았습니다..!! 감사합니다 ㅠㅠ! 이제까지는 직접 공백 수를 맞춰가며 작성했었는데 더 편해질 것 같습니다 ㅎㅎ

Rok93 replied:
IDE의 단축키들은 많이 알아두시면 생산성에 많은 도움이 됩니다. 

예를들어 지금처럼 코드 자동정렬 기능이 대표적이고 그 외에도 불필요한 Import 문들을 제거해준다거나 메서드나 클래스가 호출되고있는 위치들을 보는 기능들도 있어요.

Rok93 replied:
IDE의 단축키들은 많이 알아두시면 생산성에 많은 도움이 됩니다. 

예를들어 지금처럼 코드 자동정렬 기능이 대표적이고 그 외에도 불필요한 Import 문들을 제거해준다거나 메서드나 클래스가 호출되고있는 위치들을 보는 기능들도 있어요.

Rok93 commented on src/main/java/gift/domain/wish/service/WishService.java:
```diff
@@ -0,0 +1,66 @@
+    public void updateWish(Long id, WishUpdateRequest req, String accessToken) {
+        Member member = authService.getMemberByToken(accessToken).orElseThrow(()->new TokenExpiredException(accessToken));
+        Wish wish = wishRepository.get(id, member.getId()).orElseThrow(() -> new WishNotFoundException(id + "가 존재하지 않습니다."));
+        if (req.quantity() == 0){
```
위에서 잠시 언급했지만, 수정 API는 단순히 수정만하는게 좋을 것 같아요. 
0은 유효하지 않은 숫자이니 수정을 못하도록 막으면 더 깔끔해지지 않을까 싶어요. 😃

milk-stone replied:
넵!! 피드백 반영했습니다.

Rok93 commented on src/main/java/gift/domain/wish/repository/WishRepository.java:
```diff
@@ -0,0 +1,61 @@
+                .optional();
+    }
+
+    public int delete(Long id, Long memberId){
```
Repository는 단순히 삭제를 성공했는지 아닌지 정보만 알려주면되고 외부에서는 단순히 그 결과만 알면되지 않을까요? 

update된 row 수를 반환하기보다는 row 수가 1인지 아닌지의 결과로 반환해주는게 Repository만 알면되는 세부사항에 대해 캡슐화가 가능할 것 같아요.  (적고보니 update도 비슷한 상황이라고 생각해요 😃)

milk-stone replied:
역시 어느 계층에서든 캡슐화는 정말 중요한 것 같습니다. 좋은 통찰을 주셔서 감사합니다. 피드백 반영하겠습니다!

Rok93 commented on src/main/java/gift/domain/wish/repository/WishRepository.java:
```diff
@@ -0,0 +1,61 @@
+                .update();
+    }
+
+    public Optional<Wish> get(Long id, Long memberId) {
```
get은 너무 모호한 네이밍이라고 생각되는데요. 
조금 더 행위를 드러낼 수 있는 네이밍은 없을까요? 🤔

milk-stone replied:
Wish 객체를 받는 것이니 getWish()로 수정해보도록 하겠습니다! 확실히 뒤에 엔티티 이름 하나만 붙였을 뿐인데도 해석이 더 잘되는 느낌이네요...

Rok93 replied:
이 부분은 제가 좀 과한걸 수도 있는데 저는 가능한 메서드에 행위가 드러나는 것이 좋다고 생각하고 또 `getXXX`와 같은 네이밍은 getter와 오해의 여지가 있다고 생각하는 편이여서 getXXX, setXXX와 같은 네이밍은 지양하는 편이에요! 

이 부분은 가볍게 참고만해주세요. 🤗

Rok93 commented on src/main/java/gift/domain/wish/service/WishService.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    public void createWish(WishRequest wishRequest, String accessToken) {
+        Member member = authService.getMemberByToken(accessToken).orElseThrow(() -> new TokenExpiredException(accessToken));
```
지금처럼 로그인한 회원의 정보를 가지고와서 처리를 하는 부분을 매번 header에서 token을 가지고 온 뒤 직접 처리하고있는데요. 

[3단계 미션 소개 페이지 최하단](https://edu.nextstep.camp/s/UH6H8Iba/ls/heKNJlyS)에서 힌트로 제공하는 `HandlerMethodArgumentResolver`를 적용해보면 좋을 것 같아요.

milk-stone replied:
넵 알겠습니다! 적용해보겠습니다.

Rok93 reviewed (changes requested):
우석님 3단계도 잘 구현해주셨습니다.
몇몇 코멘트 남겨두었는데, 확인 부탁드릴게요. 🙏

milk-stone committed:
feat: 위시리스트 업데이트 시 1 이상의 값만 받도록 수정

milk-stone committed:
refactor: 함수명 직관적이게 수정, affectedRow에 대한 정보를 캡슐화, 수정 수량이 0이 되지 않도록 수정했기에 수정 API 로직에 삭제 부분 제거

milk-stone committed:
feat : HandlerMethodArgumentResolver 를 통해 컨트롤러 메서드에 진입하기 전처리를 통해 객체를 주입하도록 설정.

milk-stone committed:
feat : 2가지 기능 구현

1. @LoginMember Member member 을 통해 인증된 멤버 객체를 파라미터로 받게 하고, 그에 따라 중복되었던 멤버 인증 로직 삭제

2. quantity 값 1 이상으로 제한, Controller 단에 @Valid 태그 추가

Rok93 commented on src/main/java/gift/domain/resolver/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,48 @@
+@Component
+public class LoginMemberArgumentResolver implements HandlerMethodArgumentResolver {
+
+    private final JwtProvider jwtProvider; // 실제로는 JwtTokenProvider 등 이름 사용
```
이 주석은 무슨 용도로 작성하신건지 잘 모르겠지만 불필요한 주석은 제거해주시면 좋을 것 같아요 .

Rok93 reviewed (approved):
우석님 마지막까지 피드백 반영 잘해주셨습니다. 👍👍 
위시 리스트 미션은 이정도면 충분한 것 같아서 이만 종료하도록하겠습니다. 🎉🎉 
시간적 여유가 없으셨을텐데도 끝까지 미션 진행해주시는 모습이 인상적이었습니다. 😉
고생많으셨고 남은 미션도 화이팅입니다. 💪💪
